### ISSUE TITLE
Large thread count/lock contention

### ISSUE THREAD

**Issue by jubarat** [StackExchange/StackExchange.Redis#1006]
Our service talks to multiple Redis services using client version 2.0.513. On some of the servers we experienced an increase of thread count from 250 to 1000 accompanied by CPU increase from 60% to 100% in 3 hours time frame. Inspecting the stack trace of threads from dump shows that almost all threads were waiting for lock in `PhysicalBridge.WriteMessageTakingWriteLock` method:

```
[GCFrame: 0000003422ff9378] 
[GCFrame: 0000003422ff9508] 
[HelperMethodFrame_1OBJ: 0000003422ff9548] System.Threading.Monitor.ReliableEnterTimeout(System.Object, Int32, Boolean ByRef)
StackExchange.Redis.PhysicalBridge.WriteMessageTakingWriteLock(StackExchange.Redis.PhysicalConnection, StackExchange.Redis.Message)
StackExchange.Redis.PhysicalBridge.TryWrite(StackExchange.Redis.Message, Boolean)
StackExchange.Redis.ConnectionMultiplexer.TryPushMessageToBridge[[StackExchange.Redis.RedisValue, StackExchange.Redis]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, StackExchange.Redis.ResultBox`1, StackExchange.Redis.ServerEndPoint ByRef)
StackExchange.Redis.ConnectionMultiplexer.ExecuteAsyncImpl[[StackExchange.Redis.RedisValue, StackExchange.Redis]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, System.Object, StackExchange.Redis.ServerEndPoint)
StackExchange.Redis.RedisBase.ExecuteAsync[[StackExchange.Redis.RedisValue, StackExchange.Redis]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, StackExchange.Redis.ServerEndPoint)
StackExchange.Redis.RedisDatabase.StringGetAsync(StackExchange.Redis.RedisKey, StackExchange.Redis.CommandFlags)
... (Service code stack omited)
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
System.Threading.Tasks.Task.FinishContinuations()
System.Threading.Tasks.Task`1[[System.__Canon, mscorlib]].TrySetResult(System.__Canon)
System.Threading.Tasks.TaskCompletionSource`1[[System.__Canon, mscorlib]].TrySetResult(System.__Canon)
StackExchange.Redis.ResultBox`1[[System.__Canon, mscorlib]].TryComplete(Boolean)
StackExchange.Redis.Message.TryComplete(Boolean)
StackExchange.Redis.CompletionManager.AnyOrderCompletionHandler(System.Object)
Pipelines.Sockets.Unofficial.DedicatedThreadPoolPipeScheduler.Execute(System.Action`1, System.Object)
Pipelines.Sockets.Unofficial.DedicatedThreadPoolPipeScheduler.RunSingleItem()
System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
System.Threading.ThreadPoolWorkQueue.Dispatch()
[DebuggerU2MCatchHandlerFrame: 0000003422fff338] 

```
The threads (as noted above we use multiple Redises) which were inside lock had such a stack trace:

```
[GCFrame: 00000034c58791b8] 
[HelperMethodFrame_1OBJ: 00000034c5879288] System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object)
System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task.Wait(Int32, System.Threading.CancellationToken)
StackExchange.Redis.PhysicalConnection.FlushSync(Boolean)
StackExchange.Redis.PhysicalBridge.WriteMessageTakingWriteLock(StackExchange.Redis.PhysicalConnection, StackExchange.Redis.Message)
StackExchange.Redis.PhysicalBridge.TryWrite(StackExchange.Redis.Message, Boolean)
StackExchange.Redis.ConnectionMultiplexer.TryPushMessageToBridge[[System.Boolean, mscorlib]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, StackExchange.Redis.ResultBox`1, StackExchange.Redis.ServerEndPoint ByRef)
StackExchange.Redis.ConnectionMultiplexer.ExecuteAsyncImpl[[System.Boolean, mscorlib]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, System.Object, StackExchange.Redis.ServerEndPoint)
StackExchange.Redis.RedisBase.ExecuteAsync[[System.Boolean, mscorlib]](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1, StackExchange.Redis.ServerEndPoint)
....

```
Is it possible that the high contention around that lock was caused by `FlushSync()` taking too much time ?

**Comment by dev-day** [comment]
Hi,
Did you find the cause of this?
We see high volumes of WriteMessageTakingWriteLock for both StringGet & StringSet.
Any help would be appreciated.
Dev

**Comment by mgravell** [comment]
This is very odd. I will try to add something to track what the active
thread (the one with the lock) is doing.

**Comment by jeremyellul** [comment]
Hi, we're also experiencing high thread contention on our production environments from version 2.0.495 onwards.

Steps to reproduce, create a console app which creates 1000 threads, each thread awaits two tasks, a readasync and a writeasync operation. The read and write operations are reading and writing 80K keys to a hashset for twenty times.

No thread contention is recorded on version 1.2.6 or previous, as opposed to any version from 2.0.495 onwards.

**Comment by mgravell** [comment]
Hi, just to update: this is most likely the sync-over-async problem that I wrote about here: https://blog.marcgravell.com/2019/02/fun-with-spiral-of-death.html

The code has been fixed and this should (crosses fingers) be fixed in the upcoming deploy. I hope to be dogfooding internally today.

**Comment by guohuang** [comment]
@mgravell what exactly did you do to fix the issue? we are experiencing high thread count when using multiplexer in a fully async/await .net core application. (causing redis timeout errors)

using one connection per redis call reduced number of threads used by 50%.

**Comment by Rodrigo-Andrade** [comment]
how big are your keys/values?

**Comment by guohuang** [comment]
they are pretty small, less than 1k

**Comment by mgravell** [comment]
The combined improvements in 2.0.588 and 2.0.571 should, we believe, address a wide range of timeout and performance problems. I'm closing all timeout issues as "believed fixed in 2.0.588"; if you still see problems, please report them as new issues so we can start investigation with fresh eyes.
